smithy-model/src/main/java/software/amazon/smithy/model/transform/CreateDedicatedInputAndOutput.java

```diff
@@ -0,0 +1,220 @@
+    }
+
+    private boolean isDedicatedHeuristic(
+            OperationShape operation,
+            StructureShape struct,
+            NeighborProvider reverse,
+            RelationshipType expected
+    ) {
+        // Only assume that a shape is dedicated to the operation its name starts with the operation name.
+        if (!struct.getId().getName().startsWith(operation.getId().getName())) {
+            return false;
+        }
+
+        // Check if the shape is only referenced as input or output.
+        return isSingularReference(reverse, struct, expected);
+    }
+
+    private boolean isSingularReference(NeighborProvider reverse, Shape shape, RelationshipType expected) {
+        int totalDirectedEdges = 0;
+
+        // We need to exclude inverted edges like MEMBER_CONTAINER, and only look for directed
+        // edges like INPUT and OUTPUT.
+        for (Relationship rel : reverse.getNeighbors(shape)) {
+            if (rel.getRelationshipType().getDirection() == RelationshipDirection.DIRECTED) {
+                totalDirectedEdges++;
+                if (rel.getRelationshipType() != expected) {
+                    return false;
+                }
+            }
+        }
+
+        return totalDirectedEdges == 1;
+    }
+
+    private static StructureShape createSyntheticShape(
+            Model model,
+            OperationShape operation,
+            String suffix,
+            StructureShape source,
+            Trait inputOutputTrait
+    ) {
+        ShapeId newId = createSyntheticShapeId(model, operation, suffix);
+        return source.toBuilder()
+                .id(newId)
+                .addTrait(inputOutputTrait)
+                .build();
+    }
+
+    private static ShapeId createSyntheticShapeId(
+            Model model,
+            OperationShape operation,
+            String suffix
+    ) {
+        // Synthesize an input shape as a dedicated copy of the existing input.
+        ShapeId newId = ShapeId.fromParts(operation.getId().getNamespace(),
+                                          operation.getId().getName() + suffix);
+
+        if (model.getShapeIds().contains(newId)) {
+            ShapeId deconflicted = resolveConflict(newId, suffix);
+            if (model.getShapeIds().contains(deconflicted)) {
+                throw new ModelTransformException(String.format(
+                        "Unable to generate a synthetic %s shape for the %s operation. The %s shape already exists "
+                        + "in the model, and the conflict resolver also returned a shape ID that already exists: %s",
+                        suffix,
+                        operation.getId(),
+                        newId,
+                        deconflicted));
+            }
+            newId = deconflicted;
+        }
+
+        return newId;
+    }
+
+    private static ShapeId resolveConflict(ShapeId id, String suffix) {
+        // Say GetFooRequest exists. This then returns GetFooOperationRequest.
```

Comment:
Why change deconflict onto the input / output name rather than renaming the shape that conflicts? I ask this because doing it this way means that the input / output shapes don't have consistent names.

Reply:
The shape name that conflicts is arguably more important to have a consistent name. For example, they might appear in things like document types in the future, whereas the input/output traits would not allow the input/output shapes to appear there. But also… you could say the same thing about renaming inner shapes instead of input/output shapes— why rename those, and what strategy would you use to deconflict? These renames will be rare too.